File: problems/leetcode/0300/362/362.py
# 362. Design Hit Counter
# https://leetcode.com/problems/design-hit-counter/description/

import bisect

# Binary Search
# hits: hit가 일어났을 때의 timestamp를 저장한 배열

class HitCounter:

    def __init__(self):
        self.hits = []

    def hit(self, timestamp: int) -> None:
        self.hits.append(timestamp)

    def getHits(self, timestamp: int) -> int:
        return len(self.hits) - bisect.bisect_left(self.hits, timestamp - 299)


# Your HitCounter object will be instantiated and called as such:
# obj = HitCounter()
# obj.hit(timestamp)
# param_2 = obj.getHits(timestamp)

File: problems/leetcode/0300/362/362.cpp
// 362. Design Hit Counter
// https://leetcode.com/problems/design-hit-counter/description/

using namespace std;

#include <map>

class HitCounter {
public:
  const int INTERVAL = 299;
  map<int, int> counts;

  HitCounter() {}

  void hit(int timestamp) {
    if (counts.empty()) {
      counts[timestamp] = 1;
    } else if (counts.find(timestamp) != counts.end()) {
      counts[timestamp]++;
    } else {
      counts[timestamp] = counts.rbegin()->second + 1;
    }
  }

  int getHits(int timestamp) {
    int target = timestamp - INTERVAL;
    map<int, int>::iterator iter = counts.lower_bound(target);

    if (iter == counts.end()) {
      return 0;
    } else if (iter == counts.begin()) {
      return counts.rbegin()->second;
    }
    return (counts.rbegin()->second - prev(iter)->second);
  }
};

/**
 * Your HitCounter object will be instantiated and called as such:
 * HitCounter* obj = new HitCounter();
 * obj->hit(timestamp);
 * int param_2 = obj->getHits(timestamp);
 */
